library/psa_crypto.c

```diff
@@ -2519,39 +2521,41 @@ psa_status_t psa_hash_abort( psa_hash_operation_t *operation )
              * in use. It's ok to call abort on such an object, and there's
              * nothing to do. */
             break;
-#if defined(MBEDTLS_MD2_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD2)
         case PSA_ALG_MD2:
             mbedtls_md2_free( &operation->ctx.md2 );
             break;
 #endif
-#if defined(MBEDTLS_MD4_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD4)
         case PSA_ALG_MD4:
             mbedtls_md4_free( &operation->ctx.md4 );
             break;
 #endif
-#if defined(MBEDTLS_MD5_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD5)
         case PSA_ALG_MD5:
             mbedtls_md5_free( &operation->ctx.md5 );
             break;
 #endif
-#if defined(MBEDTLS_RIPEMD160_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_RIPEMD160)
         case PSA_ALG_RIPEMD160:
             mbedtls_ripemd160_free( &operation->ctx.ripemd160 );
             break;
 #endif
-#if defined(MBEDTLS_SHA1_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_1)
         case PSA_ALG_SHA_1:
             mbedtls_sha1_free( &operation->ctx.sha1 );
             break;
 #endif
-#if defined(MBEDTLS_SHA256_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_256)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_224)
         case PSA_ALG_SHA_224:
+#endif
         case PSA_ALG_SHA_256:
             mbedtls_sha256_free( &operation->ctx.sha256 );
             break;
 #endif
-#if defined(MBEDTLS_SHA512_C)
-#if !defined(MBEDTLS_SHA512_NO_SHA384)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_512)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_384)
         case PSA_ALG_SHA_384:
 #endif
         case PSA_ALG_SHA_512:
```

Comment:
The nesting is wrong: if `MBEDTLS_PSA_BUILTIN_ALG_SHA_384` is defined but not `MBEDTLS_PSA_BUILTIN_ALG_SHA_512`, SHA-384 won't be supported. This configuration is actually impossible right now, but it's in the backlog.

Technically the code should be
```
￼#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_384) || defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_512)
#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_384)
￼        case PSA_ALG_SHA_384:
￼#endif
￼#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_512)
￼        case PSA_ALG_SHA_512:
#endif
            mbedtls_sha512_free( &operation->ctx.sha512 );
            break;
#endif
```
but in the interest of simplicity I'd prefer this which is only extremely marginally wasteful (maybe one or two extra instructions in the compiled code if only one of the two algorithms is enabled):
```
￼#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_384) || defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_512)
￼        case PSA_ALG_SHA_384:
￼        case PSA_ALG_SHA_512:
            mbedtls_sha512_free( &operation->ctx.sha512 );
            break;
#endif
```

Likewise for SHA-224/SHA-256, and in the other functions with the same idiom.